src/stores/AutoRageshakeStore.ts

```diff
@@ -0,0 +1,122 @@
+const RAGESHAKE_INTERVAL = 5*60*1000;
+// Event type for to-device messages requesting sender auto-rageshakes
+const AUTO_RS_REQUEST = "im.vector.auto_rs_request";
+
+interface IState {
+    reportedSessionIds: Set<string>;
+    lastRageshakeTime: number;
+}
+
+/**
+ * Watches for decryption errors to auto-report if the relevant lab is
+ * enabled, and keeps track of session IDs that have already been
+ * reported.
+ */
+export default class AutoRageshakeStore extends AsyncStoreWithClient<IState> {
+    private static internalInstance = new AutoRageshakeStore();
+
+    private constructor() {
+        super(defaultDispatcher, {
+            reportedSessionIds: new Set<string>(),
+            lastRageshakeTime: 0,
+        });
+        this.onDecryptionAttempt = this.onDecryptionAttempt.bind(this);
+        this.onDeviceMessage = this.onDeviceMessage.bind(this);
+    }
+
+    public static get instance(): AutoRageshakeStore {
+        return AutoRageshakeStore.internalInstance;
+    }
+
+    protected async onAction(payload: ActionPayload) {
+        // we don't actually do anything here
+    }
+
+    protected async onReady() {
+        if (!SettingsStore.getValue("automaticDecryptionErrorReporting")) return;
+
+        if (this.matrixClient) {
+            this.matrixClient.on('Event.decrypted', this.onDecryptionAttempt);
+            this.matrixClient.on('toDeviceEvent', this.onDeviceMessage);
+        }
+    }
+
+    protected async onNotReady() {
+        if (this.matrixClient) {
+            this.matrixClient.removeListener('toDeviceEvent', this.onDeviceMessage);
+            this.matrixClient.removeListener('Event.decrypted', this.onDecryptionAttempt);
+        }
+    }
+
+    private async onDecryptionAttempt(ev: MatrixEvent): Promise<void> {
+        const wireContent = ev.getWireContent();
+        const sessionId = wireContent.session_id;
+        if (ev.isDecryptionFailure() && !this.state.reportedSessionIds.has(sessionId)) {
+            const newReportedSessionIds = new Set(this.state.reportedSessionIds);
+            await this.updateState({ reportedSessionIds: newReportedSessionIds.add(sessionId) });
+            const messageContent = {
+                "event_id": ev.getId(),
+                "room_id": ev.getRoomId(),
+                "session_id": sessionId,
+                "device_id": wireContent.device_id,
+                "user_id": ev.getSender(),
+                "sender_key": wireContent.sender_key,
+            };
+            this.matrixClient.sendToDevice(
+                AUTO_RS_REQUEST,
+                { [messageContent.user_id]: { [messageContent.device_id]: messageContent } },
+            );
+
+            const now = new Date().getTime();
+            if (now - this.state.lastRageshakeTime > RAGESHAKE_INTERVAL) {
+                await this.updateState({ lastRageshakeTime: now });
+                await sendBugReport(SdkConfig.get().bug_report_endpoint_url, {
```

Comment:
The AUTO_RS_REQUEST toDevice should contains the URL of this bug report in the `matching_issue` field